src/helpers/createTimeoutController.ts

```diff
@@ -19,6 +20,10 @@ function createTimeoutController(timeout: WaitOptions['timeout']) {
             timeoutCallbacks.forEach((callback) => callback())
             resolve()
           }, timeout)
+
+          if (jestFakeTimersAreEnabled()) {
+            jest.advanceTimersByTime(timeout - 1)
```

Comment:
I'm not sure if this implementation will work in all cases.  Consider if a condition is set to resolve just after the configured timeout.  For example overall timeout of 100ms and 50ms interval checks and check passes at 101ms:

```ts
    test('should reject if timeout is passed close to when promise resolves', async () => {
      const { waitFor } = renderHook(() => null)

      let actual = 0
      const expected = 1

      setTimeout(() => {
        actual = expected
      }, 101)

      let complete = false

      await expect(
        waitFor(
          () => {
            expect(actual).toBe(expected)
            complete = true
          },
          { timeout: 100, interval: 50 }
        )
      ).rejects.toThrow(Error('Timed out in waitFor after 100ms.'))

      expect(complete).toBe(false)
    })
```

Reply:
I see, in this case the sudo code will be 

```js
  if(jestFakeTimersAreEnabled){
  while(intervalSignal not finish){
      jest.advanceTimersByTime(interval)
      if(totalTime > Timeout){
           timeoutController.timedOut = true
      }
  }
}

```

But I am not sure how to implement it now

Reply:
The problem has fixed and test has been added

Reply:
I've been playing around with this a bit I think part of the reason it's an issue is that both the `interval` and the `timeout` controllers are advancing time, so for a 100ms `timeout` and a 50ms `interval` there is a total of 197ms (`(100-1) + 2 * (50-1)`) advanced (might be 3x the interval, because it's not quite enough to timeout with just 2).  The other issue is the way the promises work vs the timeouts where the `.then` fires after the timeout , regardless of whether it would have happened first with a real timer (I think because it's in the microtask queue but the timeout fires synchronously when the timers are advanced, but it's getting deeper than my JS knowledge goes, so I'm guessing).

I've managed to get something that can pass all the existing async tests, except one, but with a fake timers:

```
        if (allowFakeTimers && jestFakeTimersAreEnabled()) {
          const advanceTime = async (currentMs: number) => {
            if (currentMs < timeout) {
              jest.advanceTimersByTime(1)
  
              await Promise.resolve()
  
              if (finished) {
                return
              }
              await advanceTime(currentMs + 1)
            }
          }

          advanceTime(0)
        }
```

`allowFakeTimers` gets passed into `createTimoutController` and is `true` for the `interval` controller and `typeof interval !== 'number'` for the `timeout` controller (so that time still advances when interval checks are disabled, e.g. `waitForNextUpdate`).  `finished` is a `boolean` that gets set before the `timeoutController.cancel()` in the `.finally` on the end of the promise chain.

The one test that fails is failing because it has both `interval` and `timeout` disabled so there are no timeout controllers to actually advance timers, but the test is still trying to rely on `waitForNextUpdate` rather than advancing the timer itself, so part of me feels like it's an invalid scenario anyway.

Reply:
Note: I had typed this out in a stale browser and haven't looked at the more recent changes.  I'll leave the comment for you to digest, but I'll also take a look later and it might not be relevant anymore.

Reply:
That is a great Idea

The PR has been updated.

I revised the advanceTime to 
```js
const advanceTime = async () => {
    jest.advanceTimersByTime(timeout as number)
    await Promise.resolve()
  }
```

I have a question why do we need to use `await Promise.resolve()` here

and I also move the `DEFAULT_TIMEOUT` to createTimeoutController file